rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,929 @@
+    if (!result_tuple) {
+      return PyErr_Format(PyExc_RuntimeError, "Failed to create result tuple for action server");
+    }
+    return result_tuple;
+  } else {
+    const char * entity_name = PyCapsule_GetName(pyentity);
+    if (!entity_name) {
+      return NULL;
+    }
+    return PyErr_Format(PyExc_RuntimeError, "'%s' is not a known entity", entity_name);
+  }
+}
+
+#define OPTIONS_COPY_QOS_PROFILE(Options, Profile) \
+  { \
+    void * p = PyCapsule_GetPointer(py ## Profile, "rmw_qos_profile_t"); \
+    if (!p) { \
+      return NULL; \
+    } \
+    rmw_qos_profile_t * qos_profile = (rmw_qos_profile_t *)p; \
+    Options.Profile = * qos_profile; \
+    PyMem_Free(p); \
+    if (PyCapsule_SetPointer(py ## Profile, Py_None)) { \
+      /* exception set by PyCapsule_SetPointer */ \
+      return NULL; \
+    } \
+  }
+
+/// Create an action client.
+/**
+ * This function will create an action client for the given action name.
+ * This client will use the typesupport defined in the action module
+ * provided as pyaction_type to send messages over the wire.
+ *
+ * On a successful call a capsule referencing the created rcl_action_client_t structure
+ * is returned.
+ *
+ * Raises ValueError if action name is invalid
+ * Raises RuntimeError if the action client could not be created.
+ *
+ * \remark Call rclpy_action_destroy_entity() to destroy an action client.
+ * \param[in] pynode Capsule pointing to the node to add the action client to.
+ * \param[in] pyaction_type Action module associated with the action client.
+ * \param[in] pyaction_name Python object containing the action name.
+ * \param[in] pygoal_service_qos Capsule pointing to a rmw_qos_profile_t object
+ *   for the goal service.
+ * \param[in] pyresult_service_qos Capsule pointing to a rmw_qos_profile_t object
+ *   for the result service.
+ * \param[in] pycancel_service_qos Capsule pointing to a rmw_qos_profile_t object
+ *   for the cancel service.
+ * \param[in] pyfeedback_qos Capsule pointing to a rmw_qos_profile_t object
+ *   for the feedback subscriber.
+ * \param[in] pystatus_qos Capsule pointing to a rmw_qos_profile_t object for the
+ *   status subscriber.
+ * \return Capsule named 'rcl_action_client_t', or
+ * \return NULL on failure.
+ */
+static PyObject *
+rclpy_action_create_client(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pynode;
+  PyObject * pyaction_type;
+  PyObject * pyaction_name;
+  PyObject * pygoal_service_qos;
+  PyObject * pyresult_service_qos;
+  PyObject * pycancel_service_qos;
+  PyObject * pyfeedback_topic_qos;
+  PyObject * pystatus_topic_qos;
+
+  int parse_tuple_result = PyArg_ParseTuple(
+    args,
+    "OOOOOOOO",
+    &pynode,
+    &pyaction_type,
+    &pyaction_name,
+    &pygoal_service_qos,
+    &pyresult_service_qos,
+    &pycancel_service_qos,
+    &pyfeedback_topic_qos,
+    &pystatus_topic_qos);
+
+  if (!parse_tuple_result) {
+    return NULL;
+  }
+
+  if (!PyUnicode_Check(pyaction_name)) {
```

Comment:
No need for this check, `PyUnicode_AsUTF8` [calls `PyUnicode_Check`](https://github.com/python/cpython/blob/ea446409cd5f1364beafd5e5255da6799993f285/Objects/unicodeobject.c#L3870)

Reply:
Removed 34db026